### PR TITLE
fix 0-dim sum backward

### LINKED ISSUE #1152: Error on Sum.backward with zero-dimensional input
F.sum raises an error on backward when the input is zero-dimensional, since the code does not consider this case.

Code:
```python
import numpy
import chainer
import chainer.functions as F

x = chainer.Variable(numpy.array(0, dtype='f'))
y = F.sum(x)
y.backward()
# => IndexError: too many indices for array
#    (at chainer/functions/math/sum.py:50)
```

### PR DESCRIPTION
**Pull request by fukatani**
I'm going to fix #1152 at June 21st http://chainer.connpass.com/event/33526/.

### CODE REVIEW

**Comment by fukatani** [comment]
I implemented backward of sum for zero dimension and added test.

File: chainer/functions/math/sum.py [hunk 1]
```diff
@@ -46,7 +46,9 @@ def backward(self, x, gy):
         xp = cuda.get_array_module(*x)
 
         gx = xp.empty_like(x[0])
-        if self.axis is None:
+        if not gx.shape: # zero dimension check
```
Comment (okuta) [inline chainer/functions/math/sum.py]: How about `gx.ndim == 0`?

**Comment by okuta** [comment]
LGTM except comment.

**Comment by fukatani** [comment]
Thanks! `gx.ndim == 0` is more readable. In this case, comment is no need.

File: tests/chainer_tests/functions_tests/math_tests/test_sum.py [hunk 2]
```diff
@@ -98,6 +98,12 @@ def test_backward_cpu(self):
         self.check_backward(self.x, self.gy)
 
     @condition.retry(3)
+    def test_backward_zerodim_cpu(self):
+        x = numpy.array(-1, 1, ()).astype(numpy.float32)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/math_tests/test_sum.py]: Is `random.uniform` collect?

File: tests/chainer_tests/functions_tests/math_tests/test_sum.py [hunk 3]
```diff
@@ -135,6 +141,13 @@ def test_backward_gpu(self):
 
     @attr.gpu
     @condition.retry(3)
+    def test_backward_zerodim_gpu(self):
+        x = numpy.array(-1, 1, ()).astype(numpy.float32)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/math_tests/test_sum.py]: Is `random.uniform` collect?

**Comment by okuta** [comment]
Thank you!

**Pull request merged**